### PR TITLE
Implement changes to stdatomic suggested by Iain regarding default memory orders for operations where a memory order isn't valid

### PR DESCRIPTION
**Pull request by rikkimax**
As asked by @ibuclaw.

Feel free to version such defaults per compiler and per target.

### CODE REVIEW

File: druntime/src/core/stdc/stdatomic.d [hunk 1]
```diff
@@ -169,9 +169,10 @@ void atomic_flag_clear_explicit_impl()(atomic_flag* obj, memory_order order)
             break;
 
         case memory_order.memory_order_acquire:
-            // What GCC does here is generate a relaxed order (no barriers)
-            // we'll copy it, although the D compiler should really be helping out at compile time.
-            atomicStore!(memory_order.memory_order_relaxed)(&obj.b, false);
+            // Iain requested this to default to seq_cst instead of relaxed,
```
Comment (ibuclaw) [inline druntime/src/core/stdc/stdatomic.d:172]: I don't think this line is necessary. Just start with "Ideally..." and be done with it.

At worst, use @ibuclaw handle so people from the future know who you are talking about.

**Comment by rikkimax** [comment]
Azure pipelines (both): ``TARGET FAILED: compilable\testdefines.d`` again.

**Comment by thewilsonator** [comment]
> Azure pipelines (both): TARGET FAILED: compilable\testdefines.d again.

Has now been fixed.

**Review by thewilsonator — APPROVED** [review]

**Comment by ibuclaw** [comment]
Merging, as these are obvious violations owned by this module. As for all other violations it's probably going to be simpler to improve the GNU implementation of internal atomics than change ~200 lines of broken logic here. So stdatomic is kept as a dumb wrapper.

**Comment by kinke** [comment]
Hitting LLVM assertions for a number of invalid failure orderings in this module here when trying to compile its unittests...

**Comment by rikkimax** [comment]
Yes, that is to be expected.

Dmd almost totally ignores it. Iain has changed GDC to use a fallback in such a case instead of erroring.

Because it is ISA specific, and there isn't a general way of testing this in ``core.internal.atomic`` there isn't much I can do to help you here unfortunately.

I don't know how you'd solve this without ISA knowledge, or turning stdatomic into intrinsics.

**Comment by rikkimax** [comment]
Okay, I can version them to something else, you only need to turn on the version.

You pick what you want it to be replaced with.

**Comment by kinke** [comment]
How about an `assert(0)` or so for these invalid orderings? I see these 2 mentioned as invalid on cppreference too: https://en.cppreference.com/w/c/atomic/atomic_compare_exchange
> fail 	- 	the memory synchronization ordering for the load operation if the comparison fails. Cannot be [memory_order_release](https://en.cppreference.com/w/c/atomic/memory_order) or [memory_order_acq_rel](https://en.cppreference.com/w/c/atomic/memory_order) and cannot specify stronger ordering than succ

**Comment by rikkimax** [comment]
Works for me.

**Pull request merged**